src/plonk_recursion.rs

```diff
@@ -170,67 +220,147 @@ fn verify_all_ipas<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+        .all_opening_sets()
+        .iter()
+        .map(|opening_set| {
+            reduce_with_coefficients(builder, &opening_set.to_vec(), &actual_scalars)
+        })
         .collect();
 
     // Then, we reduce the above opening set reductions to a single value.
     let reduced_opening = reduce_with_powers_recursive(builder, &opening_set_reductions, v);
 
-    verify_ipa::<C, InnerC>(builder, proof, c_reduction, reduced_opening, x, ipa_challenges, security_bits)
+    let generator_n = builder.constant_wire(C::ScalarField::primitive_root_of_unity(
+        ipa_challenges.len(),
+    ));
+    let mut points: Vec<Target> = (0..2 * num_public_input_gates)
+        .step_by(2)
+        .map(|i| builder.exp_constant_usize(generator_n, i))
+        .collect();
+    points.extend(&[zeta, builder.mul(zeta, generator_n), {
+        let g65 = builder.exp_constant_usize(generator_n, GRID_WIDTH);
+        builder.mul(zeta, g65)
+    }]);
+    let halo_bs = points
+        .into_iter()
+        .map(|p| halo_g_recursive(builder, p, &ipa_challenges))
+        .collect::<Vec<_>>();
+    let halo_b = reduce_with_powers_recursive(builder, &halo_bs, v);
+
+    verify_ipa::<C, InnerC>(
+        builder,
+        proof,
+        c_reduction,
+        reduced_opening,
+        halo_b,
+        u_scaling,
+        ipa_challenges,
+        schnorr_challenge,
+        security_bits,
+    )
 }
 
 /// Verify the final IPA. Return `(u_l, u_r)`, which roughly correspond to `u` and `u^{-1}` in the
 /// Halo paper, respectively.
-fn verify_ipa<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+fn verify_ipa<C: HaloCurve, InnerC: HaloCurve<BaseField = C::ScalarField>>(
     builder: &mut CircuitBuilder<C>,
     proof: &ProofTarget,
-    p: AffinePointTarget,
-    c: Target,
-    x: Target,
+    commitment: AffinePointTarget,
+    value: Target,
+    halo_b: Target,
+    u_scaling: Target,
     ipa_challenges: Vec<Target>,
-    security_bits: usize
+    schnorr_challenge: Target,
+    security_bits: usize,
 ) -> (Vec<Target>, Vec<Target>) {
+    // The H point used for blinding polynomial commitments. Set to H(degree).
+    let pedersen_h = builder.constant_affine_point(hash_usize_to_curve::<InnerC>(
+        1 << ipa_challenges.len(),
+        security_bits,
+    ));
     // Now we begin IPA verification by computing P' and u' as in Protocol 1 of Bulletproofs.
     // In Protocol 1 we compute u' = [x] u, but we leverage to endomorphism, instead computing
     // u' = [n(x)] u.
-    // u is set to H(degree + 1). 
-    let u = builder.constant_affine_point(
-        hash_usize_to_curve::<InnerC>((1 << ipa_challenges.len()) + 1 , security_bits)
-    );
-    let u_prime = builder.curve_mul_endo::<InnerC>(CurveMulOp { scalar: x, point: u }).mul_result;
+    // u is set to H(degree + 1).
+    let u = builder.constant_affine_point(hash_usize_to_curve::<InnerC>(
+        (1 << ipa_challenges.len()) + 1,
+        security_bits,
+    ));
+    let u_prime = builder
```

Comment:
Oh yeah, seems like the recursive and non-recursive code diverges here, since the recursive code computes u' with the endo. Maybe add a TODO to fix later?

Reply:
Ah yes you're right. What do you think is best here, changing this one to `curve_mul` or changing the out-of-circuit stuff to use the endo scalar multiplication?